addon/-private/class/modifier.ts

```diff
@@ -60,27 +66,87 @@ export const _implementsLegacyHooks = <S>(
  * update if any of those values change.
  */
 export default class ClassBasedModifier<S = DefaultSignature> {
+  // Done this way with the weird combination of `declare` and `defineProperty`
+  // so that subclasses which are overriding this by writing their own `args`
+  // field type declarations continue to type check correctly. (If we introduced
+  // a getter here, existing classes defining their args via a `declare args:`
+  // would stop type checking, because TS -- correctly! -- differentiates
+  // between class fields and getters).
   /**
    * The arguments passed to the modifier. `args.positional` is an array of
    * positional arguments, and `args.named` is an object containing the named
    * arguments.
+   *
+   * @deprecated Until 4.0. Access positional and named arguments directly in
+   *   the `modify` hook instead.
    */
-  readonly args: ArgsFor<S>;
+  declare readonly args: ArgsFor<S>;
 
+  // Done this way with the weird combination of `declare` and `defineProperty`
+  // so that subclasses which are overriding this by writing their own `element`
+  // field declarations continue to type check correctly.
   /**
    * The element the modifier is applied to.
    *
    * @warning `element` is ***not*** available during `constructor` or
    *   `willDestroy`.
+   * @deprecated Until 4.0. Access the `element` as an argument in the `modify`
+   *   hook instead.
    */
-  // SAFETY: this is managed correctly by the class-based modifier. It is not
-  // available during the `constructor`.
-  // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  element: ElementFor<S> = null as any;
+  declare element: ElementFor<S>;
 
   constructor(owner: unknown, args: ArgsFor<S>) {
     setOwner(this, owner);
-    this.args = args;
+
+    // SAFETY: the point here is (for the period where we are providing `args`
+    // and `element`) to provide an internal-only way of setting and update the
+    // `args` for the modifier instance; we use the `InternalClassBasedModifier`
+    // interface to represent the internal-only API in a way that end users do
+    // *not* have access to when subclassing `ClassBasedModifier`.
+    (this as unknown as InternalClassBasedModifier<S>)[Args] = args;
+
+    // TODO: remove both of these `defineProperty`s at v4.
+    Object.defineProperty(this, 'args', {
+      enumerable: true,
+      get(this: InternalClassBasedModifier<S>) {
+        deprecate(
+          'ember-modifier: using `this.args` is deprecated. Access positional and named arguments directly in the `modify` hook instead.',
```

Comment:
I think we should try to avoid issuing this deprecation when we are **also** issuing deprecations for lifecycle methods. By definition 100% of folks using the legacy hooks are gonna _also_ get this deprecation but that secondary / tertiary deprecation doesn't really provide any additional benefit (other than flooding console).

Also, these deprecations need to include **something** to help indicate _which_ modifier is triggering the deprecation. As it stands, you could get 3x deprecations from **every** modifier use in your app, with _no_ indication of which ones are wrong....

Reply:
Following up here after some conversations over Discord and also on #220: ultimately, we *do* need to deprecate these separately from each other and from the lifecycle hooks, because we don't have a type-safe path to *not* including them when we have `modify` only, and so folks *can* use `this.args` or `this.element` when in a `modify` hook, and *can* use only one or the other (though yes, that would be shenanigans).